Create test for extracting a MALI subdomain from a larger domain

matthewhoffman opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

This merge adds a test case for a subdomain extractor.  It allows a regional domain to be quickly created from an existing whole-ice-sheet domain using a mask.  The regional mask can come from a region_mask MPAS netCDF file or a geojson file.  Optionally, ancillary files (e.g. forcing or parameter files) can be remapped at the same time.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [x] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

matthewhoffman commented:
# Testing

I created an Amery subdomain with the following cfg file on perlmutter:
```
[subdomain]

# path to file from which to extract subdomain
source_file = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/initial_conditions/AIS_4to20km_20230105/AIS_4to20km_r01_20220907_m5_drop_bed_20m_bulldoze_troughs_75_to_400m_Enderby_maxstiffness_0.8_TG_pinning_40maf_bedmap2_surface_ASE_05perc_seafloor_mu.nc

# path to region mask file for source_file
region_mask_file = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/initial_conditions/AIS_4to20km_20230105/AIS_4to20km_r01_20220907.regionMask_ismip6.nc

# region number to extract
region_number = 3

# filename for the subdomain to be generated
dest_file_name = Amery.nc

# mesh projection to be used for setting lat/lon values
# Should match the projection used in the source_file
# Likely one of 'ais-bedmap2' or 'gis-gimp'
mesh_projection = ais-bedmap2

# whether to extend mesh into the open ocean along the ice-shelf margin
# This is necessary if the region mask ends right along the ice-shelf edge,
# or if the ice-shelf is covered by two regions.
# It is recommended to try extracting a subdomain with this False, and if the
# ocean buffer is inadequate, trying again with this True.
extend_ocean_buffer = True

# number of iterations to grow open ocean buffer along ice-shelf margin
# Only used if extend_ocean_buffer is True
# Should be equal to approximately the number of ocean buffer cells in the
# source_file
grow_iters = 15
```

Result:
![image](https://user-images.githubusercontent.com/4182034/224750468-35260d7c-acf5-40a3-adc2-38822a7623c2.png)

Result setting  `extend_ocean_buffer = False`:
![image](https://user-images.githubusercontent.com/4182034/224750693-c7a129a4-e670-4621-86fc-8fed4c54f319.png)

matthewhoffman commented:
Temporary docs are here: https://portal.nersc.gov/cfs/fanssie/hoffman2/compass-docs-subdomain-extractor/html/index.html
Specifically:
* https://portal.nersc.gov/cfs/fanssie/hoffman2/compass-docs-subdomain-extractor/html/users_guide/landice/test_groups/mesh_modifications.html
* https://portal.nersc.gov/cfs/fanssie/hoffman2/compass-docs-subdomain-extractor/html/developers_guide/landice/test_groups/mesh_modifications.html
* https://portal.nersc.gov/cfs/fanssie/hoffman2/compass-docs-subdomain-extractor/html/developers_guide/landice/api.html#mesh-modifications

trhille commented:
@matthewhoffman, is this expected behavior for `extend_mesh = False` (left) vs `True` (right)? i.e., is it always desirable to extend only into floating ice and exclude ice rises, etc?
![image](https://user-images.githubusercontent.com/17446278/230631446-600a5dab-a782-4420-b619-d661d8154312.png)

trhille commented on compass/landice/tests/mesh_modifications/subdomain_extractor/extract_region.py:
```diff
@@ -0,0 +1,176 @@
+        # (otherwise not necessary to save to disk)
+        write_netcdf(dsMaskOut, 'cull_mask.nc')
+
+        # cull the mesh
```
Note that I'm planning to refactor the mesh test cases soon, which will probably move a lot of these mesh operations (down to ~L158) to `compass.landice.mesh` to reduce repetition between test cases.

trhille reviewed (commented):
@matthewhoffman, thanks for creating this great tool! See my comments regarding a small amount of repeated code for the flood-fill, and also my question about the desired behavior of `extend_mesh`. I'll wait to approve until you respond to those.

matthewhoffman commented on compass/landice/tests/mesh_modifications/subdomain_extractor/extract_region.py:
```diff
@@ -0,0 +1,176 @@
+            conc = dsMesh['cellsOnCell'][:].values
+            neonc = dsMesh['nEdgesOnCell'][:].values
+
+            # First grow forward to capture any adjacent ice shelf
```
done in 3174f0e17

matthewhoffman commented on compass/landice/tests/mesh_modifications/subdomain_extractor/extract_region.py:
```diff
@@ -0,0 +1,176 @@
+
+            print('Starting ocean grow fill')
+            # find cells at current ocean bdy
+            for iter in range(grow_iters):
```
done in 3174f0e17

matthewhoffman commented on compass/landice/tests/mesh_modifications/subdomain_extractor/subdomain_extractor.cfg:
```diff
@@ -0,0 +1,31 @@
+region_mask_file = TO BE SUPPLIED BY USER
+
+# region number to extract
+region_number = 1
```
I think that might overcomplicate the test case for what I imagine is a fairly rare use case.  The test case takes just a couple minutes to run, so even if wanting to process a handful of regions, it should not take too much time to run each region serially.

However, I did add the ability to use a geojson file to specify a custom region.  I kept running into situations where the existing regions had in adequate gutter and the gutter extrapolation wasn't able to produce good results.  This would also be useful for creating larger regions.

matthewhoffman commented:
@trhille , I addressed the comments from your review and added a few significant new features:
* ability to use geojson file instead of existing region_mask file
* ability to also remap ancillary files on the same mesh (e.g. forcing, parameter files)
* the second item required adding a new remap method using ncremap instead of our standard MALI interpolation tools
* I also moved the flood fill to landice framework

On Perlmutter this cfg can be used to test this PR out.  As set up, setting `region_definition = region_mask_file` will generate a subdomain mesh and 3 forcing files for Amery.  Changing that to `region_definition = geojson` will instead create a subdomain for our Thwaites Glacier 'minimal' region definition.

```
[subdomain]

# path to file from which to extract subdomain
source_file = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/initial_conditions/AIS_4to20km_20230105/AIS_4to20km_r01_20220907_m5_drop_bed_20m_bulldoze_troughs_75_to_400m_Enderby_maxstiffness_0.8_TG_pinning_40maf_bedmap2_surface_ASE_05perc_seafloor_mu.nc

# method for defining region
# one of 'region_mask_file' or 'geojson'
region_definition = geojson
#region_definition = region_mask_file

# path to geojson file to be used if region_definition='geojson'
geojson_file = /global/cfs/cdirs/fanssie/users/hoffman2/compass/subdomain_extractor_cleanup/compass/landice/tests/thwaites/thwaites_minimal.geojson


# path to region mask file for source_file
region_mask_file = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/initial_conditions/AIS_4to20km_20230105/AIS_4to20km_r01_20220907.regionMask_ismip6.nc

# region number to extract
region_number = 3

# filename for the subdomain to be generated
dest_file_name = Subdomain.nc

# mesh projection to be used for setting lat/lon values
# Should match the projection used in the source_file
# Likely one of 'ais-bedmap2' or 'gis-gimp'
mesh_projection = ais-bedmap2

# whether to extend mesh into the open ocean along the ice-shelf margin
# This is necessary if the region mask ends right along the ice-shelf edge,
# or if the ice-shelf is covered by two regions.
# It is recommended to try extracting a subdomain with this False, and if the
# ocean buffer is inadequate, trying again with this True.
extend_ocean_buffer = True

# number of iterations to grow open ocean buffer along ice-shelf margin
# Only used if extend_ocean_buffer is True
# Should be equal to approximately the number of ocean buffer cells in the
# source_file
grow_iters = 15

# method for performing interpolation
# 'ncremap' uses pyremap to call ESMF_RegridWeightGen to generate a
# nstd weight file and then uses ncremap to perform remapping.
# This method supports interpolating ancillary files (below)
# but likely needs to be run on a compute node and is more fragile.
# 'mali_interp' uses the MALI interpolation script interpolate_to_mpasli_grid.py
# This method does not support ancillary files but may be more robust
# and can likely be run quickly on a login node.
interp_method = ncremap

# optional forcing files that could also be interpolated
# interpolating these files requires using the 'ncremap' interp_method
extra_file1 = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/forcing/ais_mesh_4to20km_res/tier1_experiments/expAE01/AIS_4to20km_r01_20220907_smb_NorESM1-M_RCP26-repeat_2300_minus1_bare_land.nc
extra_file2 = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/forcing/ais_mesh_4to20km_res/tier1_experiments/ctrlAE/AIS_4to20km_r01_20220907_obs_TF_1995-2017_8km_x_60m.nc
extra_file3 = /global/cfs/cdirs/fanssie/MALI_projects/ISMIP6-2300/initial_conditions/AIS_4to20km_20230105/basin_and_coeff_gamma0_DeltaT_quadratic_non_local_after_relaxation.nc
extra_file4 = None
extra_file5 = None
```

trhille reviewed (approved):
Nice work, @matthewhoffman! I'm approving based on inspection and your testing. Someone else should test independently, as we discussed.

xylar commented on docs/developers_guide/landice/test_groups/mesh_modifications.rst:
```diff
@@ -0,0 +1,58 @@
+.. _mesh_modifications:
```
@matthewhoffman, this should have been named `dev_mesh_modifications`.  It has the same name as the section in the user's guide, which is producing warnings when building the documentation.  Could you open a small PR to fix this when you have time?

(I accidentally commented previously on the commit, not the PR.)

matthewhoffman replied:
Addressed in #808
